### PR TITLE
[Sema] [CSDiag] Fix incorrect diagnostic when using keypath with tuples

### PR DESCRIPTION
**Pull request by theblixguy**
This PR fixes an issue where an incorrect diagnostic would be emitted when using key path with tuples. 

Resolves [SR-9213](https://bugs.swift.org/browse/SR-9213).

### CODE REVIEW

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -6685,7 +6685,12 @@ static bool diagnoseKeyPathComponents(ConstraintSystem &CS, KeyPathExpr *KPE,
                                            : defaultUnqualifiedLookupOptions),
                                corrections);
 
-      if (currentType)
+      if (currentType->getKind() == TypeKind::Tuple) {
```
[the remark this answers is not in the document]
Comment (theblixguy) [reply lib/Sema/CSDiag.cpp]: Oh, you're right, `currentType` could be null! I assumed the `rootType` of the `KeyPathExpr` won't be null (if it was then the existing incorrect diagnostic wouldn't be emitted), but there's no harm in checking though!

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -510,6 +510,8 @@ ERROR(expr_string_interpolation_outside_string,none,
       "string interpolation can only appear inside a string literal", ())
 ERROR(unsupported_keypath_tuple_element_reference,none,
       "key path cannot reference tuple elements", ())
+ERROR(expr_swift_keypath_unimplemented_tuple,none,
```
[the remark this answers is not in the document]
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def]: Removed.

File: test/expr/unary/keypath/keypath-unimplemented.swift [hunk 3]
```diff
@@ -12,3 +12,13 @@ class C {
...
+
+struct TupleKeypath {
+    let labeled: (foo: Int, bar: String)
+    let unlabeled: (Int, Int)
+}
+
+let _: KeyPath<TupleKeypath, Int> = \TupleKeypath.labeled.foo // expected-error {{key path support for tuples is not implemented}}
+let _: KeyPath<TupleKeypath, String> = \TupleKeypath.labeled.bar // expected-error {{key path support for tuples is not implemented}}
+let _: KeyPath<TupleKeypath, Int> = \TupleKeypath.unlabeled.0 // expected-error {{key path support for tuples is not implemented}}
+let _: KeyPath<TupleKeypath, String> = \TupleKeypath.unlabeled.1 // expected-error {{key path support for tuples is not implemented}}
```
Comment (xedin) [inline test/expr/unary/keypath/keypath-unimplemented.swift]: Could you please a newline at the end of this file?

File: lib/Sema/CSDiag.cpp [hunk 4]
```diff
@@ -6686,8 +6686,14 @@ static bool diagnoseKeyPathComponents(ConstraintSystem &CS, KeyPathExpr *KPE,
                                corrections);
 
       if (currentType)
-        TC.diagnose(componentNameLoc, diag::could_not_find_type_member,
-                    currentType, componentName);
+        if (currentType->is<TupleType>()) {
```
Comment (xedin) [inline lib/Sema/CSDiag.cpp]: Just FYI but you can do `isa<TupleType>(currentType.getPointer())` e.g. that's the only check for presence of `ParenType`.

**Review by xedin — APPROVED** [review]
LGTM

**Comment by xedin** [comment]
@theblixguy once you fix this small newline nit I mentioned, I'll run the CI for you.

File: test/expr/unary/keypath/keypath-unimplemented.swift (hunk 3 above)
Comment (theblixguy) [reply test/expr/unary/keypath/keypath-unimplemented.swift]: Done!

File: lib/Sema/CSDiag.cpp (hunk 4 above)
Comment (theblixguy) [reply lib/Sema/CSDiag.cpp]: Oh, thanks for the tip! 😄

**Comment by theblixguy** [comment]
> @theblixguy once you fix this small newline nit I mentioned, I'll run the CI for you.

Done! @xedin

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@xedin Seems like it can be merged now 🎉

**Comment by xedin** [comment]
@theblixguy Thank you!

File: lib/Sema/CSDiag.cpp [hunk 5]
```diff
@@ -6686,8 +6686,14 @@ static bool diagnoseKeyPathComponents(ConstraintSystem &CS, KeyPathExpr *KPE,
                                corrections);
 
       if (currentType)
-        TC.diagnose(componentNameLoc, diag::could_not_find_type_member,
-                    currentType, componentName);
+        if (isa<TupleType>(currentType.getPointer())) {
```
Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: Aha, this still isn't quite the solution you want: it doesn't look through typealiases.

```
typealias X = (Int, Int)

struct Foo {
  let x: X
}

let kp = \Foo.x.1
```

Instead, you can use `currentType->is<TupleType>()`, which will look through typealiases and other forms of "sugar".

**Review by jrose-apple — COMMENTED** [review]
Thanks for doing this!

File: lib/Sema/CSDiag.cpp (hunk 5 above)
Comment (theblixguy) [reply lib/Sema/CSDiag.cpp]: That's what I did initially, but @xedin said I do `isa<TupleType>(currentType.getPointer())`. I'll create a new PR to fix this!

Comment (xedin) [reply lib/Sema/CSDiag.cpp]: Sorry if my comment was misleading, I didn’t mean you have to use `isa`, but rather that it’s a one way to check without skipping sugar...

Comment (theblixguy) [reply lib/Sema/CSDiag.cpp]: Ah okay, I understand now 😄 Thanks for clarifying @xedin!

Comment (theblixguy) [reply lib/Sema/CSDiag.cpp]: Okay :) @xedin

**Pull request merged**